### PR TITLE
New architecture

### PR DESCRIPTION
**Pull request by nkcr**
Moves the code around to split it into packages. Here is a global overview of how it looks now:

global architecture

### CODE REVIEW

**Comment by nkcr** [comment]
Sorry for the too-long PR. It should be the last for the Go part, as the architecture is now getting to a nice shape. Here is what remains after that (expect subsequent PRs ✌️):

- [ ] Update comments and docs
- [ ] Use constants for the object#action strings
- [ ] (if time permits) add tests

**Comment by pierluca** [comment]
Re: your last comment: no need to add tons of tests, but a few examples would be good... we will later improve the test coverage with the students.

**Comment by nkcr** [comment]
> Re: your last comment: no need to add tons of tests, but a few examples would be good... we will later improve the test coverage with the students.

Yes that's the plan, but I'd like to close that PR first, and then also do the things I promised to do when we closed the other PRs ;)

File: be1-go/channel/lao/mod.go [hunk 1]
```diff
@@ -0,0 +1,738 @@
...
+
+		if db != nil {
+			log.Printf("inserting attendee into db")
+
+			err := insertAttendee(db, attendee, c.channelID)
+			if err != nil {
+				log.Printf("error: failed to insert attendee into db: %v", err)
+			}
+		}
+	}
```
Comment (pierluca) [inline be1-go/channel/lao/mod.go:585]: same thing, I'd love (some) of it to just be embedded within the Inbox and hidden away from the LAO channel.

**Review by pierluca — CHANGES_REQUESTED** [review]
Couple of odd design decisions, but overalls this looks absolutely great !

File: be1-go/channel/election/mod.go [hunk 2]
```diff
@@ -171,26 +168,12 @@ func (c *electionChannel) Publish(publish method.Publish) error {
...
 		case "cast_vote":
-			var castVote messagedata.VoteCastVote
-
-			err := msg.UnmarshalData(&castVote)
-			if err != nil {
-				return xerrors.Errorf("failed to unmarshal cast vote: %v", err)
-			}
-
-			err = c.castVoteHelper(msg, castVote)
+			err := c.castVoteHelper(msg)
```
[the remark this answers is not in the document]
Comment (nkcr) [reply be1-go/channel/election/mod.go:171]: True. This PR addresses the architecture though, not the weird naming students came with 😅.

File: be1-go/channel/inbox/mod.go [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+
+// AddWitnessSignature adds a signature of witness to a message of ID
+// `messageID`. if the signature was correctly added return true otherwise
+// returns false
+func (i *Inbox) AddWitnessSignature(messageID string, public string, signature string) error {
+	msg, ok := i.GetMessage(messageID)
+	if !ok {
+		// TODO: We received a witness signature before the message itself. We
+		// ignore it for now but it might be worth keeping it until we actually
+		// receive the message
```
[the remark this answers is not in the document]
Comment (nkcr) [reply be1-go/channel/inbox/mod.go:56]: I'm not sure what to do with that. Do you want this feature to be implemented in this PR ?

File: be1-go/channel/inbox/mod.go [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+
+	msg.WitnessSignatures = append(msg.WitnessSignatures, message.WitnessSignature{
+		Witness:   public,
+		Signature: signature,
+	})
+
+	if os.Getenv("HUB_DB") != "" {
```
[the remark this answers is not in the document]
Comment (nkcr) [reply be1-go/channel/inbox/mod.go:69]: Absolutely true. The plan when implementing the DB was to eventually use a storage abstraction. In the meantime yes we could gather all those operation in the inbox. It was something difficult to do with the old architecture but now it comes out pretty obviously :)

File: be1-go/channel/lao/mod.go (hunk 1 above)
Comment (nkcr) [reply be1-go/channel/lao/mod.go:585]: Thing is the attendees are stored on the hub (organizer) and not in the inbox. Thus the need to have db handling in both places.

File: be1-go/channel/election/mod.go [hunk 5]
```diff
@@ -1,35 +1,41 @@
-package hub
+package election
```
[the remark this answers is not in the document]
Comment (nkcr) [reply be1-go/channel/election/mod.go:1]: To be ~fought~ discussed 🕊

**Comment by nkcr** [comment]
As discussed, instead of refactoring the storage I moved the database file retrieval in the same place so that we already have something usable for the student. Though it does not remove the plan to have a better storage implementation.

**Review by pierluca — APPROVED** [review]
LGTM

**Pull request merged**